Sources/AWSLambdaDeploymentDescriptor/DeploymentDescriptor.swift

```diff
@@ -0,0 +1,132 @@
+// ===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftAWSLambdaRuntime open source project
+//
+// Copyright (c) 2022 Apple Inc. and the SwiftAWSLambdaRuntime project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftAWSLambdaRuntime project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// ===----------------------------------------------------------------------===//
+
+import Foundation
+import Yams
+
+// this is the developer-visible part of the deployment decsriptor.
+// the rest is generated automatically.
+public protocol DeploymentDescriptor {
+    
+    // returns the event sources for a given Lambda function
+    func eventSources(_ lambdaName: String) -> [EventSource]
+    
+    // returns environment variables to associate with the given Lambda function
+    func environmentVariables(_ lambdaName: String) -> EnvironmentVariable
+
+    // returns additional resources to create in the cloud (ex : a dynamoDB table)
+    func addResource() -> [Resource]
+}
+
+// Generates a deployment descriptor modelized by DeploymentDefinition
+extension DeploymentDescriptor {
+    
+    // create the SAM deployment descriptor data structure
+    // this method use multiple data sources :
+    // - it calls protocol functions implemented by the lambda function developer to get function-specific
+    //   details (event source, environment variables)
+    // - it uses command line arguments (list of lambda names)
+    // - it uses some default values (like the ZIP file produced by archive command)
+    public func deploymentDefinition() -> DeploymentDefinition {
+        
+        // Create function resources for each Lambda function
+        var resources = self.lambdaNames().map { name in
+            return createServerlessFunctionResource(name) // returns [Resource]
+        }.flatMap{ $0 } // convert [[Resource]] to [Resource]
+            
+        // add developer-provided resources
+        resources.append(contentsOf: self.addResource())
+
+        // create the SAM deployment descriptor
+        return SAMDeployment(resources: resources)
+    }
+    
+    // returns environment variables to associate with the given Lambda function
+    // This is a default implementation to avoid forcing the lambda developer to implement it when not needed
+    public func environmentVariables(_ lambdaName: String) -> EnvironmentVariable {
+        return EnvironmentVariable.none()
+    }
+    
+    // returns additional resources to create in the cloud (ex : a dynamoDB table)
+    // This is a default implementation to avoid forcing the lambda developer to implement it when not needed
+    public func addResource() -> [Resource] {
+        return Resource.none()
+    }
+    
+    // entry point and main function. Lambda developer must call this function.
+    public func run() {
```

Comment:
can we use an @main based protocol instead?

Reply:
Agreed. This is not needed anymore with the `Deploy.swift` declarative approach